Sdesk 7469

jerome-poisson opened the pull request:
This PR is still a WIP. The classes are quite extensive, and the methods calling each other in a layered fashion make it challenging to follow. I have several questions about how to best manage certain method or type porting that I would like to tackle.

jerome-poisson committed:
[ARCHIVE] Add async `archive` and `archived` resources and services.

SDESK-7469

jerome-poisson committed:
Fix `Resource.rel` FIXME:

- use `validate_data_relation_async` to convert legacy `Resource.rel`.
- add `LinkedInPackage` dataclass instead of generic dict.

MarkLark86 commented on superdesk/archive_async/module.py:
```diff
@@ -0,0 +1,50 @@
+        indexes=[
+            MongoIndexOptions(
+                name="uri_1",
+                keys=[("uri", 1), ("background", 1)],
```
[unique](https://github.com/superdesk/superdesk-core/blob/async/superdesk/core/types/mongo.py#L35) defaults to `True`, therefor you would need to add `unique=False` all these index options

MarkLark86 commented on superdesk/archive_async/module.py:
```diff
@@ -0,0 +1,50 @@
+            ),
+        ],
+    ),
+    rest_endpoints=RestEndpointConfig(
```
We will continue to use the old resource for the REST API, therefor we won't need `rest_endpoints` config here for now (this will be done at a later stage)

MarkLark86 commented on superdesk/archive_async/service.py:
```diff
@@ -0,0 +1,77 @@
+    mediaService = ArchiveMediaService()
+    cropService = CropService()
+
+    async def on_fetched(self, docs) -> None:
```
These `on_fetched` don't belong on the new AsyncResourceService as they have nothing to do with the data layer (these are used on REST APIs). They can be removed from this service (will be added to a custom REST API class later for this resource)

MarkLark86 commented on superdesk/archive_async/service.py:
```diff
@@ -0,0 +1,77 @@
+        for item in items:
+            handle_existing_data(item)
+
+    async def on_create(self, docs: list[ArchiveResourceModel]) -> None:
```
Is there a way to make a base class that this and the original resource service can both use? The old Eve resources now support async `on_` functions (such as `def on_create`, `def on_update` etc), just not on the core method functions (such as `def create`, `def update` etc).

I've done this with the async publish endpoints to reduce the amount of code changes required. I used the [package service](https://github.com/superdesk/superdesk-core/blob/develop/apps/packages/package_service.py#L93) as an example (pure python class, implementing the `on_create` etc methods).

And then you would update both the old and this service to use those functions.
Keeping in mind, if the class makes changes to the dictionary (especially here in creation), then you would need to propagate those changes onto the Pydantic model instance), maybe just using the [clone_with](https://github.com/superdesk/superdesk-core/blob/async/superdesk/core/types/model.py#L127) function

MarkLark86 commented on superdesk/archived_async/module.py:
```diff
@@ -0,0 +1,34 @@
+        indexes=[
+            MongoIndexOptions(
+                name="item_id_1",
+                keys=[("item_id", 1), ("background", 1)],
```
Same here, will require `unique=False`

MarkLark86 commented on superdesk/archived_async/module.py:
```diff
@@ -0,0 +1,34 @@
+            ),
+        ],
+    ),
+    rest_endpoints=RestEndpointConfig(
```
And here as well, REST config not required (front-end will continue to use the old REST API for now)

MarkLark86 commented on superdesk/types/archive.py:
```diff
@@ -0,0 +1,453 @@
+# This file is part of Superdesk
```
We also have the [Content API resource model](https://github.com/superdesk/superdesk-core/blob/async/content_api/items/model.py). There might be a lot of common fields here.

The ContentAPI items are slightly different to archive items, and it is also used by Newshub. But hopefully you might be able to combine them (and maybe move some of those classes into a `superdesk.types` module for common dataclasses etc

MarkLark86 commented on superdesk/types/archive.py:
```diff
@@ -0,0 +1,453 @@
+    sub_label: str
+
+
+@dataclass
```
We have this defined as a core field, [Geopoint](https://github.com/superdesk/superdesk-core/blob/async/superdesk/core/resources/fields.py#L209), which provides elastic mapping, json schema and validation

MarkLark86 commented on superdesk/types/archive.py:
```diff
@@ -0,0 +1,453 @@
+
+class MetadataResource(ResourceModel):
+    guid: str
+    uri: str
```
Any string fields in the old resource that have `"mapping": not_analyzed"` need to use [fields.Keyword](https://github.com/superdesk/superdesk-core/blob/async/superdesk/core/resources/fields.py#L209), that way the elastic mapping will be correct

MarkLark86 replied:
There is also `TextWithKeyword` and `HTML` fields defined in core fields

MarkLark86 commented on superdesk/types/archive.py:
```diff
@@ -0,0 +1,453 @@
+class MetadataResource(ResourceModel):
+    guid: str
+    uri: str
+    unique_id: int
```
Any fields in the old resource that has `unique` validation, requires [validate_unique_value_async](https://github.com/superdesk/superdesk-core/blob/async/superdesk/core/resources/validators.py#L204) from core validators.

MarkLark86 commented on superdesk/types/archive.py:
```diff
@@ -0,0 +1,453 @@
+    original_source: str
+    ingest_provider_sequence: str
+    usageterms: str | None = None
+    copyrightnotice: str | None = None
```
Any fields in the old resource that has `"mapping": not_indexed` will need to be annotated with [fields.not_indexed()](https://github.com/superdesk/superdesk-core/blob/async/superdesk/core/resources/fields.py#L205) so Elastic mapping will be correct.

MarkLark86 commented on superdesk/types/archive.py:
```diff
@@ -0,0 +1,453 @@
+    copyrightnotice: str | None = None
+    copyrightholder: str | None = None
+    anpa_category: list[Subject] | None = None
+    subject: list[Subject]
```
Any list fields that has `"type": "nested"` in the original schema needs to be [annotated](https://github.com/superdesk/superdesk-core/blob/async/superdesk/core/resources/fields.py#L182) as such here as well, so the elastic mapping is correct.

In this case it would be:
```python
subject: Annotated[list[Subject] | None, fields.nested_list(include_in_parent=True)] = None
```

MarkLark86 commented on superdesk/types/archive.py:
```diff
@@ -0,0 +1,453 @@
+
+@dataclass
+class Subject:
+    qcode: str
```
Please go over all usage of `not_analyzed` in the original schema, as it's used a lot, and replace with `fields.Keyword`

MarkLark86 commented on superdesk/types/archive.py:
```diff
@@ -0,0 +1,453 @@
+    subject: list[Subject]
+    genre: list[Genre] | None = None
+    company_codes: list[CompanyCodes] | None = None
+    item_type: ContentTypes = Field(ContentTypes.TEXT, alias="type")
```
Using an enum is good, but we also need to mark it as `not_analyzed` so elastic mapping is correct. You can annotate it with [fields.keyword_mapping](https://github.com/superdesk/superdesk-core/blob/async/superdesk/core/resources/fields.py#L170) in this case (I haven't tried it with an Enum field before, but it would work)

MarkLark86 commented on superdesk/types/archive.py:
```diff
@@ -0,0 +1,453 @@
+    item_type: ContentTypes = Field(ContentTypes.TEXT, alias="type")
+    package_type: str | None = Field(None, deprecated=True)
+    language: str | None = None
+    abstract: str | None = None
```
Any field that has `"mapping": string_with_analyzer` in the old schema needs to use the [fields.HTML](https://github.com/superdesk/superdesk-core/blob/async/superdesk/core/resources/fields.py#L144) type, as it uses the same `html_field_analyzer`

MarkLark86 commented on superdesk/types/archive.py:
```diff
@@ -0,0 +1,453 @@
+    language: str | None = None
+    abstract: str | None = None
+    headline: str | None = None
+    slugline: Slugline
```
Slugline is a special case and needs custom elastic mapping. This can be achieved using the [fields.elastic_mapping](https://github.com/superdesk/superdesk-core/blob/async/superdesk/core/resources/fields.py#L166) annotation. In this case it would be:

```python
Slugline: Annotated[
    str,
    fields.elastic_mapping(
        {
            "type": "string",
            "fielddata": True,
            "fields": {
                "phrase": {
                    "type": "string",
                    "analyzer": "phrase_prefix_analyzer",
                    "fielddata": True,
                },
                "keyword": {
                    "type": "keyword",
                },
                "text": {"type": "string", "analyzer": "html_field_analyzer"},
            }
        }
    )
]
```

This should hopefully give you the same elastic mapping as the original schema

MarkLark86 commented on superdesk/types/archive.py:
```diff
@@ -0,0 +1,453 @@
+    rewritten_by: str | None = None
+    # FIXME: duplicate of ItemSchema
+    sequence: int | None = None
+    keywords: list[str]
```
You can also use fields.Keyword with lists, i.e. `list[fields.Keyword]`

MarkLark86 commented on superdesk/types/archive.py:
```diff
@@ -0,0 +1,453 @@
+    priority: int | None = None
+    urgency: int | None = None
+    profile: str | None = None
+    state: ContentStates
```
These 3 fields also need to be annotated by the fields.keyword_mapping, to make sure they aren't analyzed by Elastic

MarkLark86 commented on superdesk/types/archive.py:
```diff
@@ -0,0 +1,453 @@
+
+
+@dataclass
+class Author:
```
Have a look at the [ContentAuthor](https://github.com/superdesk/superdesk-core/blob/async/content_api/items/model.py#L74) type from the ContentAPI, might be able to be re-used here

MarkLark86 commented on superdesk/types/archive.py:
```diff
@@ -0,0 +1,453 @@
+    authors: list[Author] | None = None
+    description_text: str | None = None
+    archive_description: str | None = None
+    groups: list[GroupRef] | None = None
```
Any fields with `"minlength"` or `"maxlength"` defined in the old schema can use the [validate_minlength](https://github.com/superdesk/superdesk-core/blob/async/superdesk/core/resources/validators.py#L62) and/or [validate_maxlength](https://github.com/superdesk/superdesk-core/blob/async/superdesk/core/resources/validators.py#L87) validators from core

MarkLark86 commented on superdesk/types/archive.py:
```diff
@@ -0,0 +1,453 @@
+    country: str
+    code: str
+    scheme: str
+    location: list[Geopoint]
```
Is `location` a list or a dict, seems to be a [dict](https://github.com/superdesk/superdesk-core/blob/develop/superdesk/metadata/item.py#L141) in the original schema

MarkLark86 commented on superdesk/types/archive.py:
```diff
@@ -0,0 +1,453 @@
+    qcode: str
+    scheme: str
+    feature_class: str
+    location: list[Geopoint]
```
And again here, seems to be a dict not a list

MarkLark86 commented on superdesk/types/archive.py:
```diff
@@ -0,0 +1,453 @@
+    body_html: str | None = None
+    body_text: str | None = None
+    dateline: Dateline | None = None
+    expiry: datetime
```
Expiry may not always have a value, so we can make it optional with `datetime | None = None`

MarkLark86 commented on superdesk/types/archive.py:
```diff
@@ -0,0 +1,453 @@
+    dateline: Dateline | None = None
+    expiry: datetime
+    # FIXME: type file
+    media: str | None = None
```
Hmmm, actually not sure if we currently support `"type": "file"` in the new async lib.

This is because it's a special field, where Eve will grab the file from the HTTP request, store the file, and replace this field's value with the ID of the uploaded file.

I will need to look into adding support for this. For now we can mark this with `fields.Keyword` and place a `# TODO-ASYNC: Add support for `file` and `media` field types

MarkLark86 commented on superdesk/types/archive.py:
```diff
@@ -0,0 +1,453 @@
+    media: str | None = None
+    mimetype: str
+    poi: POI
+    renditions: dict[str, dict]
```
Just an FYI, any field with a type annotated of `dict` will automatically have `"mapping": not_enabled` functionality applied (nothing to do here, just wanting to let you know that)

MarkLark86 commented on superdesk/types/archive.py:
```diff
@@ -0,0 +1,453 @@
+
+@dataclass
+class Associations:
+    featuremedia: FeatureMedia | None = None
```
This would cause a validation error for non featuremedia associations.

Maybe we could just use the fields.elastic_mapping feature to implement proper elastic mapping, and use simple dict as the type. i.e.

```python
Associations = Annotated[
    dict,
    fields.elastic_mapping(
        "type": "object",
        "dynamic": False,
        "properties": {
            "featuremedia": {  # keep indexing featuremedia - we do some filtering using it
                "type": "object",
                "dynamic": False,
                "properties": {
                    "_id": not_analyzed,
                    "guid": not_analyzed,
                    "unique_id": {"type": "integer"},
                },
            }
        },
    )
]

jerome-poisson replied:
There is `ContentAssociation` defined at `content_api/items/model.py` which does a similar thing, I'm using it.

MarkLark86 commented on superdesk/types/archive.py:
```diff
@@ -0,0 +1,453 @@
+
+
+@dataclass
+class EntityMetadata:
```
Is this correct, I don't see the `metadata` field in the original schema. The fields that use this seem to just be `list[EntityMetadataItem]`

MarkLark86 commented on superdesk/types/archive.py:
```diff
@@ -0,0 +1,453 @@
+    organisation: EntityMetadata | None = None
+    creditline: str
+    linked_in_packages: list[LinkedInPackage] | None = None
+    highlight: Annotated[str | None, validate_data_relation_async("highlights")] = None
```
Should these use [fields.ObjectId](https://github.com/superdesk/superdesk-core/blob/async/superdesk/core/resources/fields.py#L150), so serialisation works as expected.

MarkLark86 commented on superdesk/types/archive.py:
```diff
@@ -0,0 +1,453 @@
+
+@dataclass
+class MarkedDesk:
+    desk_id: Annotated[str | None, validate_data_relation_async("desks")] = None
```
Same here, should probably use fields.ObjectId. In fact, anything that uses `Resource.rel` in the old schema, where the `type` param is not changed, should use fields.ObjectId

MarkLark86 commented on superdesk/types/archive.py:
```diff
@@ -0,0 +1,453 @@
+    flags: Flag | None = None
+    sms_message: str | None = None
+    format: str = FORMATS.HTML
+    auto_publish: bool
```
This should probably have a default of `False`

MarkLark86 commented on superdesk/types/archive.py:
```diff
@@ -0,0 +1,453 @@
+
+@dataclass
+class Attachment:
+    pass
```
Missing `attachment` field, resource link to the `attachments` resource

MarkLark86 commented on superdesk/types/archive.py:
```diff
@@ -0,0 +1,453 @@
+    publish_schedule: datetime | None = None
+    # FIXME: duplicate of ItemSchema
+    schedule_settings: ScheduleSettings | None = None
+    used: bool
```
This could probably be defaulted to `False`

MarkLark86 commented on superdesk/types/archive.py:
```diff
@@ -0,0 +1,453 @@
+    # FIXME: duplicate of ItemSchema
+    schedule_settings: ScheduleSettings | None = None
+    used: bool
+    used_count: int
```
Could probably default to `0`, what do you think?

MarkLark86 commented on superdesk/types/archive.py:
```diff
@@ -0,0 +1,453 @@
+    used_count: int
+    used_updated: datetime
+    metrics: dict[str, Any]
+    _type: str
```
This is already defined in parent resource class, no need to re-define here

MarkLark86 reviewed (commented):
@jerome-poisson I have gone over the old schema and this one. There are some Elastic mapping differences between these models and the old ones. We need to make sure that the old elastic schema and new ones match, otherwise users may have search issues.

Maybe you can grab the elastic mapping of the old one, and create a unit test for this new one, to ensure the mappings are the same. Check [here](https://github.com/superdesk/superdesk-core/blob/async/tests/core/resource_model_test.py#L75) for an example test

jerome-poisson committed:
update service

jerome-poisson committed:
Merge branch 'async' into SDESK-7469

jerome-poisson committed:
Update following review:

- disable Rest endpoint
- use a `BaseContentItem` model as parent class for `MetadataResource`
  and `ContentAPIItem`
- use `fields.Keyword` when `not_analyzed` is used in legacy resource.
- added default values for many fields
- removed some custom models in favor of existing ones from `content_api.items.model`

jerome-poisson committed:
- use `BaseContentItem` as base class for `ContentAPIItem`.

jerome-poisson committed:
more changes following review.

- removed on_fetched_item method
- adjusted many types and default values
- some fields have been removed from `BaseContentItem` to have a
  different implementation in `MetadataResource` following review,
  without changing behaviour of `ContentAPIItem`
- added validators
- make `validate_unique_value_async` work without `resource_name`
  specified, so it can be used with class inheritance.

jerome-poisson commented:
All review comments should have been addressed, there is now a base class `BaseContentItem` use for `ContentAPIItem` and `MetadataResource`.

MarkLark86 reviewed (approved):
Looks good to merge
